pkg/controller/seed-controller-manager/update/update_controller.go

```diff
@@ -116,16 +117,36 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 }
 
 func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
-
+	// If cluster is not healthy yet there is nothing to do.
+	// If it gets healthy we'll get notified by the event. No need to requeue.
 	if !cluster.Status.ExtendedHealth.AllHealthy() {
-		// Cluster not healthy yet. Nothing to do.
-		// If it gets healthy we'll get notified by the event. No need to requeue
 		return nil, nil
 	}
 
-	clusterType := v1.KubernetesClusterType
-	if cluster.IsOpenshift() {
-		clusterType = v1.OpenShiftClusterType
+	clusterType := getClusterType(cluster)
+
+	// Check if cluster has non-empty initial machine deployment request serialized
+	// into its annotations and create it if its there.
+	if request, ok := cluster.Annotations[v1.InitialMachineDeploymentRequestAnnotation]; ok && request != "" {
+		var nodeDeployment *v1.NodeDeployment
+		if err := json.Unmarshal([]byte(request), &nodeDeployment); err != nil {
+			return nil, fmt.Errorf("cannot unmarshal initial machine deployment request for %s cluster: %v", cluster.Name, err)
+		}
+
+		if !r.isInitialMachineDeploymentCreated() {
+			if err := r.createInitialMachineDeployment(nodeDeployment); err != nil {
+				return nil, fmt.Errorf("cannot create initial machine deployment for %s cluster: %v", cluster.Name, err)
+			} else {
+				r.recorder.Eventf(cluster, corev1.EventTypeNormal, "InitialMachineDeployment", "Created initial machine deployment")
+				if err := r.removeInitialMachineDeploymentRequestAnnotation(cluster); err != nil {
+					return nil, fmt.Errorf("cannot remove initial machine deployment request from %s cluster: %v", cluster.Name, err)
+				}
+			}
+		} else {
+			if err := r.removeInitialMachineDeploymentRequestAnnotation(cluster); err != nil {
+				return nil, fmt.Errorf("cannot remove initial machine deployment request from %s cluster: %v", cluster.Name, err)
+			}
+		}
```

Comment:
@xrstf @moadqassem can you take a look on the above code and tell me if this is be okay for you? I will implement methods soon but I would like to confirm that it is okay for you if the approach will be like this.

Reply:
What made you choose this controller to add your logic to? Because it deals with nodes during the cluster upgrades? What would you say to writing a dedicated controller just for this purpose (turning the initial-node annotation into a MachineDeployment)?

Reply:
I thought that this was our agreement during discussion on Slack as you agreed on "re-using an existing controllers" and @moadqassem pointed this one. It's good that the code is not finished yet and there won't be much to revert, so no worries here. However, at this point I'd rather leave it to lifecycle team as only the controller part is missing. I think it will be quicker for you to do and we will avoid any unnecessary misunderstandings.

Reply:
@xrstf What is the problem of adding the initial node deployment here too. This controller is not just responsible for taking care of control plane components, it is also responsible for managing the machine list deployment within the user cluster. Also there is no real semantics here that this controller is only for upgrades, as it watches seed resources(cluster) and in the very same time it has the reconciled user cluster client.

Reply:
Yeah it's fine.

Reply:
So I think we're good here, @moadqassem WDYT?

Reply:
SGTM